onnxscript/ir/_external_data.py

```diff
@@ -51,3 +75,209 @@ def set_base_dir(graph: _core.Graph | _core.GraphView, base_dir: str | os.PathLi
+        external_tensor = _core.ExternalTensor(
+            file_path,
+            tensor_info.offset,
+            tensor_info.length,
+            tensor.dtype,  # type: ignore[arg-type]
+            shape=tensor.shape,  # type: ignore[arg-type]
+            name=tensor.name,  # type: ignore[arg-type]
+        )
+        external_tensors.append(external_tensor)
+    return external_tensors
+
+
+def convert_tensors_to_external(
+    tensors: list[_protocols.TensorProtocol],
+    abs_path: str | os.PathLike,
+) -> list[_core.ExternalTensor]:
+    """
+    This method takes in tensors to be converted to external tensors and returns the converted external tensors.
+
+    Args:
+        tensors: Tensors to be converted to external tensors
+        abs_path: Absolute path to which external data is to be stored.
+    """
+    external_data_info: list[tuple[_protocols.TensorProtocol, ExternalDataInfo]] = []
+    # Sort all tensors based on tensor sizes, in order to avoid unneccesarry alignment.
+    # All the smaller tensors are written earlier and alignment is performed for the larger tensors.
+    sorted_indices = sorted(range(len(tensors)), key=lambda i: tensors[i].nbytes)
+    tensors = [tensors[i] for i in sorted_indices]
+
+    current_offset = 0
+    for tensor in tensors:
+        tensor_info = _set_external_data(tensor, current_offset)
+        external_data_info.append((tensor, tensor_info))
+        current_offset = tensor_info.offset + tensor_info.length
+    _save_external_data(external_data_info, abs_path)
+
+    # Convert initializers to ExternalTensors
+    external_tensors = _convert_as_external_tensors(external_data_info, abs_path)
+    # Sort external_tensors based on original key order
+    external_tensors = [
+        external_tensors[i]
+        for i in sorted(range(len(external_tensors)), key=lambda i: sorted_indices[i])
+    ]
+    return external_tensors
+
+
+def to_external_data(
+    model: _core.Model,
+    base_path: str | os.PathLike,
+    relative_path: str | os.PathLike,
+    load_external_to_memory: bool = False,
+) -> _core.Model:
+    """
+    Call to set all tensors with raw data as external data.
+
+    Args:
+        model: Model to be converted.
+        base_path: Path of base directory.
+        relative_path: Path to which external data is to be stored.
+        load_external_to_memory: If set to true, loads external tensors present in the same file path as destination path to memory. Otherwise, the external tensors are appended to file.
+    """
+    path = os.path.join(base_path, relative_path)
+    # Check if file path is valid, and create subsequent subdirectories within the path if they don't exist
+    parent_path = os.path.join(*path.split("/")[:-1])
+    tmp_path = os.path.join(base_path, "tmp")
+    os.makedirs(parent_path, exist_ok=True)
+    os.makedirs(tmp_path, exist_ok=True)
+    # Check if file is empty. If not, load pre-existing external data.
+    if os.stat(path).st_size != 0:
```

Comment:
Check existence instead? Also, if none of the tensors from the model are referring to the file, we don't even need to bother renaming and all that. So I would consider checking if any tensors are using this file first, before doing any of this caching

Reply:
What would be expected to happen in the scenario where none of the tensors reference the file however, the destination path the user is provided the same. Do we still overwrite/rename this file?